#### Contents

#### Removing Evaluation Column

#### Renaming orders_dow to orders_day_of_week

#### Changing Variable Data Type

#### Transposing Data

#### Creating Data Dictionaries

#### Subsetting Data

#### Task 4.4

## 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Removing Eval Column

In [2]:
# Making Path do data as a string
path = r'C:\Users\bbissonette\Documents\Data Analytics Course\Acheivement 4\Instacart Basket Analysis'

In [5]:
# importing orders data
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [6]:
# Dropping eval_set col from orders data set
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [8]:
# count of days_since_prior_order frequencies
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

## 03. Renaming orders_dow to orders_day_of_week

In [9]:
# renaming dow column
df_ords.rename(columns = {'order_dow':'orders_day_of_week'}, inplace = True)

In [10]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

## 04. Changing Variable Data Type

In [11]:
# changing order id from int to str
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [12]:
df_ords['order_id'].dtype

dtype('O')

## 05. Transposing Data

In [13]:
# importing and defining the department df
df_dep = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [14]:
# showing whats in the the deparment df
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [15]:
# transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [16]:
# defining new dep df
df_dep_t = df_dep.T

In [17]:

df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
# adding indect to the dep df
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [19]:
# Taking the first row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [20]:
new_header

0    department
Name: department_id, dtype: object

In [21]:
# removing first line under header
df_dep_t_new = df_dep_t[1:]

In [22]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [23]:
# setting the header row as the df header
df_dep_t_new.columns = new_header

In [24]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## 06. Data Dictionaries

In [25]:
# making a data dictionay of the department df
data_dict = df_dep_t_new.to_dict('index')

In [26]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [27]:
# importing and defining products df
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

In [28]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [29]:
# finding that 19 stands for
print(data_dict.get('19'))

{'department': 'snacks'}


## 07. Subsetting

In [31]:
# indexing the code
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [32]:
# identifying what dep id 19 shows
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [ ]:
# 1st subset code
df_snacks =  df_prods[df_prods['department_id']==19]

In [34]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [35]:
# 2nd subset code with loc function
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [36]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [37]:
# 3RD subset code
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [38]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Start Task 4.4

### Step 2

In [40]:
# discovering the data types from orders
df_ords.dtypes

order_id                   object
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [41]:
# chaning user_id from interger to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [42]:
df_ords['user_id'].dtype

dtype('O')

### Step 3

In [43]:
# changing order_hour_of_day to order_hour for simplicity
df_ords.rename(columns = {'order_hour_of_day':'order_hour'}, inplace = True)

In [44]:
df_ords.head()

order_id user_id eval_set  order_number  orders_day_of_week  order_hour  \
0  2539329       1    prior             1                   2           8   
1  2398795       1    prior             2                   3           7   
2   473747       1    prior             3                   3          12   
3  2254736       1    prior             4                   4           7   
4   431534       1    prior             5                   4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Step 4

In [48]:
# finding the busiest hour for placing order
df_ords['order_hour'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour, dtype: int64

The Highest number of Orders occurred at 10:00 AM with 288,418 orders

### Step 5

In [50]:
# finding the meaning to value of 4 withing deparment id col
df_prods[df_prods['department_id']==4]

product_id                                      product_name  aisle_id  \
30             31                                White Pearl Onions       123   
42             43                               Organic Clementines       123   
44             45                                 European Cucumber        83   
65             66                         European Style Spring Mix       123   
88             89                    Yogurt Fruit Dip Sliced Apples       123   
...           ...                                               ...       ...   
49582       49578                                Black Garlic Bulbs       123   
49623       49619                                        Opo Squash        83   
49639       49635  Baby Food Blueberry, Parsnip & Buckwheat Stage 2        83   
49661       49657                                 Cabernet Tomatoes        83   
49687       49683                                    Cucumber Kirby        83   

       department_id  prices  
30                 4     7.5  
42                 4    11.5  
44                 4    14.3  
65                 4    11.7  
88                 4    12.6  
...              ...     ...  
49582              4     8.0  
49623              4    12.7  
49639              4    12.5  
49661              4     8.3  
49687              4    13.2  

[1684 rows x 5 columns]

### Step 6

In [51]:
# finding a list of the top breakfast items
df_breakfast_items = df_prods[df_prods['department_id']==14]

In [52]:
df_breakfast_items.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

### Step 7

In [57]:
# 5, 20, 7, 12 are the dinner party elements. Trying to find the dinner party top items
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [59]:
df_dinner_parties.head(30)

product_id                                       product_name  aisle_id  \
2             3               Robust Golden Unsweetened Oolong Tea        94   
6             7                     Pure Coconut Water With Orange        98   
9            10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10           11                                  Peach Mango Juice        31   
16           17                                  Rendered Duck Fat        35   
19           20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22           23                             Organic Turkey Burgers        49   
34           35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38           39           Daily Tangerine Citrus Flavored Beverage        64   
39           40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43           44                        Sparkling Raspberry Seltzer       115   
48           49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51           52                                Mirabelle Brut Rose       134   
52           53              Whole Leaf Pure Aloe With Lemon Juice        98   
61           62            Premium Deli Oven Roasted Turkey Breast        96   
68           69                                                NaN        26   
72           73                  Jasmine Tea Unfiltered Ginger Ale        77   
73           74                    Artisan Chick'n & Apple Sausage        14   
81           82                       All Natural 100% Apple Juice        98   
83           84                                         Lamb Shank         7   
84           85                               Soppressata Piccante        96   
86           87                             Classics Earl Grey Tea        94   
96           97                        Organic Chamomile Lemon Tea        94   
108         109                             Grape Leaf Hummus Wrap        13   
118         119                             Chardonnay Paso Robles        62   
137         138                                 Olli Napoli Salami        96   
146         147                           Organic Ranch Veggie Dip        67   
149         150                                          Brut Rosé       134   
157         158  Smoked Ham & Oven Roasted Turkey Sub Sandwich Kit        96   
166         167                        Sparkling Blush Grape Juice        77   

     department_id  prices  
2                7     4.5  
6                7     4.4  
9                7     8.4  
10               7     2.8  
16              12    17.1  
19               7     6.0  
22              12     8.2  
34              12    15.1  
38               7    12.5  
39              12    22.5  
43               7    12.2  
48              20    10.1  
51               5    14.4  
52               7     4.7  
61              20    14.6  
68               7    11.8  
72               7     8.3  
73              20     3.0  
81               7     6.4  
83              12    24.3  
84              20     8.9  
86               7     9.1  
96               7     8.4  
108             20     7.1  
118              5     5.5  
137             20     5.1  
146             20     3.4  
149              5    12.9  
157             20    14.0  
166              7    13.8

### Step 8

In [61]:
# Finding the number of row from the last df made
df_dinner_parties.shape

(7650, 5)

The data frame has 7650 rows

### Step 9-10

In [63]:
# making user_id 1 into a subset then finding info on that user id
df_userid1 = df_ords[df_ords['user_id']=='1']

In [64]:
df_userid1

order_id user_id eval_set  order_number  orders_day_of_week  order_hour  \
0   2539329       1    prior             1                   2           8   
1   2398795       1    prior             2                   3           7   
2    473747       1    prior             3                   3          12   
3   2254736       1    prior             4                   4           7   
4    431534       1    prior             5                   4          15   
5   3367565       1    prior             6                   2           7   
6    550135       1    prior             7                   1           9   
7   3108588       1    prior             8                   1          14   
8   2295261       1    prior             9                   1          16   
9   2550362       1    prior            10                   4           8   
10  1187899       1    train            11                   4           8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [65]:
df_userid1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11 entries, 0 to 10
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   order_id                11 non-null     object 
 1   user_id                 11 non-null     object 
 2   eval_set                11 non-null     object 
 3   order_number            11 non-null     int64  
 4   orders_day_of_week      11 non-null     int64  
 5   order_hour              11 non-null     int64  
 6   days_since_prior_order  10 non-null     float64
dtypes: float64(1), int64(3), object(3)
memory usage: 704.0+ bytes


In [67]:
# finding stats on user id 1
df_userid1.describe()

order_number  orders_day_of_week  order_hour  days_since_prior_order
count     11.000000           11.000000   11.000000               10.000000
mean       6.000000            2.636364   10.090909               19.000000
std        3.316625            1.286291    3.477198                9.030811
min        1.000000            1.000000    7.000000                0.000000
25%        3.500000            1.500000    7.500000               14.250000
50%        6.000000            3.000000    8.000000               19.500000
75%        8.500000            4.000000   13.000000               26.250000
max       11.000000            4.000000   16.000000               30.000000

In [68]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [69]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))